src/Illuminate/Support/ServiceProvider.php

```diff
@@ -46,17 +46,34 @@ public function __construct($app)
     }
 
     /**
-     * Merge the given configuration with the existing configuration.
+     * Merge the configuration at the given path with the existing configuration.
      *
      * @param  string  $path
      * @param  string  $key
      * @return void
      */
     protected function mergeConfigFrom($path, $key)
     {
-        $config = $this->app['config']->get($key, []);
+        $this->app['config']->set(
+            $key,
+            $this->mergeConfig(
+                $key,
+                require $path,
+                $this->app['config']->get($key, [])
+            )
+        );
+    }
 
-        $this->app['config']->set($key, array_merge(require $path, $config));
+    /**
+     * Merge two configurations together.
+     * @param  string $key
+     * @param  array  $baseConfig
+     * @param  array  $mergeConfig
+     * @return array
+     */
+    protected function mergeConfig($key, array $baseConfig, array $mergeConfig)
```

Comment:
If there are multiple keys in a service provider you might need to handle each one differently. You'd be able to identify which one was being merged based on `$key`.